core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -169,20 +190,29 @@ private void subscribe0(Subscriber<? super T> subscriber, Executor executor, boo
             failLateSubscriber(subscriber, executor, AbortedStreamException.get());
             return;
         }
-
         final PendingSubscription<? super T> newPendingSubscription =
                 new PendingSubscription<>(subscriber, executor, withPooledObjects);
 
-        final StreamMessage<T> delegate = this.delegate;
-        if (delegate == null) {
-            if (!pendingSubscriptionUpdater.compareAndSet(this, null, newPendingSubscription)) {
-                failLateSubscriber(subscriber, executor,
-                                   new IllegalStateException("subscribed by other subscriber already"));
-            }
+        if (!subscriptionUpdater.compareAndSet(this, null, newPendingSubscription)) {
+            failLateSubscriber(subscriber, executor,
+                               new IllegalStateException("subscribed by other subscriber already"));
             return;
         }
 
-        subscribeToDelegate(delegate, newPendingSubscription);
+        // There are 4 state in class.
+        // If subscribe is called first, then it will match INIT -> SUBSCRIBED.
+        // If delegate is called first, than it will match DELEGATED -> SUBSCRIBED_AND_DELEGATED.
```

Comment:
- subscribe -> subscribe()
- delegate -> delegate()
- then it will match -> the state will transit from ... to ...